server/src/test/java/com/broadcom/lsp/cobol/service/ClientServerIntegrationTest.java

```diff
@@ -0,0 +1,211 @@
+    assertEquals(
+        SHUTDOWN_RESPONSE, service.formatting(new DocumentFormattingParams()).get()); // NOSONAR
+
+    List<TextDocumentContentChangeEvent> textEdits = new ArrayList<>();
+    textEdits.add(new TextDocumentContentChangeEvent(INCORRECT_TEXT_EXAMPLE));
+    service.didChange(
+        new DidChangeTextDocumentParams(
+            new VersionedTextDocumentIdentifier(UseCaseUtils.DOCUMENT_URI, 0), textEdits));
+
+    service.didClose(
+        new DidCloseTextDocumentParams(new TextDocumentIdentifier(UseCaseUtils.DOCUMENT_URI)));
+
+    ((TestLanguageServer) server).revokeShutdown();
+    assertEquals(1, ((DisposableLanguageServer) server).getExitCode());
+  }
+
+  @Test
+  void testIncorrectLanguageId() {
+
+    service.didOpen(
+        new DidOpenTextDocumentParams(
+            new TextDocumentItem(UseCaseUtils.DOCUMENT_URI, "incorrectId", 1, TEXT_EXAMPLE)));
+    UseCaseUtils.await(() -> !client.getMessagesToShow().isEmpty());
+    assertTrue(
+        client.getMessagesToShow().stream()
+            .map((Function<MessageParams, Object>) MessageParams::getMessage)
+            .anyMatch(
+                it ->
+                    it.toString()
+                        .equals(
+                            "Cannot find a language engine for the given language ID: incorrectId")));
+  }
+
+  @Test
+  void testNotAllowedFileExtensionAnalysis() {
+    service.didOpen(
+        new DidOpenTextDocumentParams(
+            new TextDocumentItem(CPY_DOCUMENT_URI, LANGUAGE, 1, TEXT_EXAMPLE)));
+    UseCaseUtils.await(() -> !client.getMessagesToLog().isEmpty());
+    assertTrue(
+        client.getMessagesToLog().stream()
+            .map((Function<MessageParams, Object>) MessageParams::getMessage)
+            .anyMatch(
+                it ->
+                    it.toString()
+                        .equals("The given document extension is unsupported: " + CPY_EXTENSION)));
+  }
+
+  @Test
+  void testFindMultipleCopybookReferences() throws ExecutionException, InterruptedException {
+    setUpCopybook();
+    List<? extends Location> locations = invokeReferencesRequest(TEST_COPYBOOK1, true);
+    assertEquals(4, locations.size());
+
+    assertContainsRange(locations, range(20, 12, COPY_LENGTH));
+    assertContainsRange(locations, range(22, 12, COPY_LENGTH));
+  }
+
+  @Test
+  void testFindSingleCopybookReference() throws ExecutionException, InterruptedException {
+    setUpCopybook();
+    List<? extends Location> locations = invokeReferencesRequest(TEST_COPYBOOK2, true);
+    assertEquals(3, locations.size());
+
+    assertContainsRange(locations, range(21, 12, COPY_LENGTH));
+  }
+
+  private void setUpCopybook() {
+    client.clean();
+    List.of(new CobolText("CPYBK1", COPYBOOK1), new CobolText("CPYBK2", COPYBOOK2)).stream()
+        .map(UseCaseUtils::toCopybookModel)
+        .forEach(copybookService::store);
+    UseCaseUtils.runTextValidation(service, TEXT);
+    UseCaseUtils.waitForDiagnostics(client);
```

Comment:
Those methods are deprecated and may be replaced by mocking the `LanguageEngineFacade`

Reply:
Have waited on the future. Not mocked as the className says integration test. But I am doubtful if we need these.